modules/server/src/main/scala/scaladex/server/service/PublishProcess.scala

```diff
@@ -65,35 +61,50 @@ class PublishProcess(
       userState: Option[UserState]
   ): Future[PublishResult] = {
     val pomRef = s"${pom.groupId}:${pom.artifactId}:${pom.version}"
-    githubExtractor.extract(pom) match {
-      case None =>
-        // TODO: save artifact with no github information
-        Future.successful(PublishResult.NoGithubRepo)
-      case Some(repo) =>
-        // userState can be empty when the request of publish is done through the scheduler
-        if (userState.isEmpty || userState.get.hasPublishingAuthority(env) || userState.get.repos.contains(repo)) {
-          converter.convert(pom, repo, creationDate) match {
-            case Some((artifact, deps)) =>
-              for {
-                isNewProject <- database.insertArtifact(artifact, deps, Instant.now)
-                _ <-
-                  if (isNewProject && userState.nonEmpty)
-                    updateGithubInfo(new GithubClient(userState.get.info.token), artifact.projectRef, Instant.now())
-                  else Future.successful(())
-              } yield {
-                logger.info(s"Published $pomRef")
-                PublishResult.Success
-              }
-            case None =>
-              logger.warn(s"Cannot convert $pomRef to valid Scala artifact.")
-              Future.successful(PublishResult.InvalidPom)
+    val maybeProjectRef = githubExtractor.extract(pom)
+    if (isPermittedToPublish(userState, maybeProjectRef)) {
+      converter.convert(pom, maybeProjectRef, creationDate) match {
+        case Some((artifact, deps)) =>
+          insertArtifactWithDependencies(artifact, deps, maybeProjectRef, userState).map { _ =>
+            logger.info(s"Published $pomRef")
+            PublishResult.Success
           }
-        } else {
-          logger.warn(s"User ${userState.get.info.login} attempted to publish to $repo")
-          Future.successful(PublishResult.Forbidden(userState.get.info.login, repo))
-        }
+        case None =>
+          logger.warn(s"Cannot convert $pomRef to a valid Scala artifact")
+          Future.successful(PublishResult.InvalidPom)
+      }
+    } else {
+      logger.warn(s"User ${userState.get.info.login} attempted to publish to $maybeProjectRef")
+      Future.successful(PublishResult.Forbidden(userState.get.info.login, maybeProjectRef))
+    }
+  }
+
+  private def insertArtifactWithDependencies(
+      artifact: Artifact,
+      dependencies: Seq[ArtifactDependency],
+      maybeProjectRef: Option[Project.Reference],
+      maybeUserState: Option[UserState]
+  ): Future[Unit] =
+    database.insertArtifact(artifact, dependencies, Instant.now).flatMap { isNewProject =>
+      if (isNewProject && maybeProjectRef.isDefined && maybeUserState.isDefined)
```

Comment:
I imagine this logic will have to change to account for some other cases, e.g.

What if `isNewProject = false` but `maybeProejctRef.isDefined = true`? In this case, we should update the GitHub information, correct?

Reply:
> What if isNewProject = false but maybeProejctRef.isDefined = true? In this case, we should update the GitHub information, correct?

Updating the GithubInfo is expensive because we have a limited rate of requests to the Github API. So here we only want to do it on new projects. For the existing projects we have a scheduled job that runs slowly in the background every day to update the GithubInfo.